core/savings.go

```diff
@@ -5,82 +5,130 @@ import (
+}
+
+func (s *Savings) CostTotal() float64 {
+	return s.costGrid + s.costSelfConsumption
+}
+
+func (s *Savings) EffectivePrice() float64 {
+	if s.ChargedTotal() == 0 {
+		return s.currentGridPrice()
+	}
+	return s.CostTotal() / s.ChargedTotal()
 }
 
-func (s *Savings) ChargedSelfConsumption() float64 {
-	return s.chargedSelfConsumption
+func (s *Savings) SavingsAmount() float64 {
+	return s.chargedSelfConsumption * (s.currentGridPrice() - s.currentFeedInPrice())
 }
 
-func (s *Savings) shareOfSelfProducedEnergy(gridPower float64, pvPower float64, batteryPower float64) float64 {
+func (s *Savings) shareOfSelfProducedEnergy(gridPower, pvPower, batteryPower float64) float64 {
 	batteryDischarge := math.Max(0, batteryPower)
 	batteryCharge := math.Min(0, batteryPower) * -1
 	pvConsumption := math.Min(pvPower, pvPower+gridPower-batteryCharge)
 
 	gridImport := math.Max(0, gridPower)
 	selfConsumption := math.Max(0, batteryDischarge+pvConsumption+batteryCharge)
 
-	selfPercentage := 100 / (gridImport + selfConsumption) * selfConsumption
+	share := selfConsumption / (gridImport + selfConsumption)
 
-	if math.IsNaN(selfPercentage) {
+	if math.IsNaN(share) {
 		return 0
 	}
 
-	return selfPercentage
+	return share
 }
 
-func (s *Savings) Update(gridPower float64, pvPower float64, batteryPower float64, chargePower float64) {
-	now := s.Clock.Now()
+func (s *Savings) currentGridPrice() float64 {
+	if s.tariffs.Grid != nil {
+		if gridPrice, err := s.tariffs.Grid.CurrentPrice(); err == nil {
+			return gridPrice
+		}
+	}
+	return 0.3
+}
 
-	selfPercentage := s.shareOfSelfProducedEnergy(gridPower, pvPower, batteryPower)
+func (s *Savings) currentFeedInPrice() float64 {
+	if s.tariffs.FeedIn != nil {
+		if gridPrice, err := s.tariffs.FeedIn.CurrentPrice(); err == nil {
+			return gridPrice
+		}
+	}
+	return 0.08
+}
+
+func (s *Savings) Update(p publisher, gridPower, pvPower, batteryPower, chargePower float64) {
+	// assume charge power as constant over the duration -> rough kWh estimate
+	addedEnergy := s.clock.Since(s.updated).Hours() * chargePower / 1e3
+	share := s.shareOfSelfProducedEnergy(gridPower, pvPower, batteryPower)
 
-	updateDuration := now.Sub(s.updated)
+	addedSelfConsumption := addedEnergy * share
+	addedGrid := addedEnergy - addedSelfConsumption
 
-	// assuming the charge power was constant over the duration -> rough estimate
-	addedEnergy := updateDuration.Hours() * chargePower / 1000
+	s.chargedGrid += addedGrid
+	s.costGrid += addedGrid * s.currentGridPrice()
+	s.chargedSelfConsumption += addedSelfConsumption
+	s.costSelfConsumption += addedSelfConsumption * s.currentFeedInPrice()
 
-	s.chargedTotal += addedEnergy
-	s.chargedSelfConsumption += addedEnergy * (selfPercentage / 100)
-	s.updated = now
+	s.updated = s.clock.Now()
 
-	s.log.DEBUG.Printf("%.1fkWh charged since %s", s.chargedTotal, time.Since(s.started).Round(time.Second))
+	s.log.DEBUG.Printf("%.1fkWh charged since %s", s.ChargedTotal(), time.Since(s.started).Round(time.Second))
```

Comment:
können wir die Zeilen noch in 1 kombinieren? Brauchen wir das jetzt nach der Debugphase _überhaupt_ noch? Weniger Loglines = _viel_ einfachere Fehleranalyse.

Mein Vorschlag: raus, Rest ist im Cache sichtbar.

Reply:
Jo, brauchen wir nicht mehr. ist raus.